### PR TITLE
Log a warning message for long running storage operations when flushing

### PR DESCRIPTION
**Pull request by Frederik-Baetens**

### CODE REVIEW

File: src/workerd/io/actor-cache.c++
Comment (Frederik-Baetens) [inline src/workerd/io/actor-cache.c++:2644]: didn't add extra scope here because I figured the tine added by the fulfill call would be minimal.

File: src/workerd/util/duration-alert.h
Comment (jqmmes) [inline src/workerd/util/duration-alert.h:17]: nit: You can use `LOG_NOSENTRY` here.

**Comment by jqmmes** [comment]
This LGTM, but I guess you need runtime approval. I find it strange that touching `actor-cache` doesn't have `Workers Durable Objects` as CODEOWNERS.

File: src/workerd/util/duration-alert.h
Comment (Frederik-Baetens) [reply src/workerd/util/duration-alert.h:17]: Hmm, I tried, but it doesn't seem to play nice with passing the logMessage as an argument rather than a string literal. Don't really know why that is. I'd assume b/c of it being a macro defined in terms of another macro perhaps?

File: src/workerd/util/duration-alert.h
Comment (jclee) [inline src/workerd/util/duration-alert.h:7]: Class names are typically in PascalCase?

File: src/workerd/io/actor-cache.c++
Comment (jclee) [inline src/workerd/io/actor-cache.c++:2586]: I'm not sure this does what you're expecting...  I suspect the compiler is allowed to instantiate a temporary "durationAlert" value for the scope of the statement, then free it before doing the `co_await`.  If you want the value to exist for the scope of the code block, I think some further action is needed, like putting it in a named variable?

I guess one way to test this would be to add a `usleep(5000 * 1000);` statement in the block and see if the log fires.  Or alternately, to put in several `KJ_DBG()` statements to verify that the destructor runs after the end of the code block.

File: src/workerd/util/duration-alert.h
Comment (jclee) [inline src/workerd/util/duration-alert.h:20]: Per the style guide, explicit destructors should typically be declared `noexcept(false)`.

https://github.com/capnproto/capnproto/blob/master/style-guide.md#exceptions-usage

File: src/workerd/util/duration-alert.h
Comment (jclee) [reply src/workerd/util/duration-alert.h:22]: Yeah, `LOG_NOSENTRY` does its concatenation by putting two string literals next to each other to make a new string literal -- e.g. `("a" "b")` -- so it would only be expected to work when the first macro argument is itself a string literal, not a variable.

I suppose it also might be worth considering using some sort of limiting mechanism like `LOG_ERROR_PERIODICALLY` does, if we expect a delay would cause a flood of these messages at once in syslog.  But maybe it's hard to judge up front how necessary that might be.

File: src/workerd/util/duration-alert.h
Comment (jclee) [inline src/workerd/util/duration-alert.h:21]: Might have meant `clock.now() - start` ?

Comment (jclee) [reply src/workerd/util/duration-alert.h:21]: If you wanted to write a `KJ_TEST` unit test to confirm this class works as expected, you might be able to use the same mechanism as the kj debug test to capture the log output (declare an instance of `ExceptionCallback` on the stack implementing `logMessage()`):

https://github.com/capnproto/capnproto/blob/958ccfe/c%2B%2B/src/kj/debug-test.c%2B%2B#L168-L175

...along with using `kj::TimerImpl::advanceTo()` to simulate advancing the clock without actually waiting, e.g.:

https://github.com/capnproto/capnproto/blob/958ccfe4/c%2B%2B/src/kj/compat/http-test.c%2B%2B#L3843

File: src/workerd/io/actor-cache.c++
Comment (Frederik-Baetens) [reply src/workerd/io/actor-cache.c++]: This indeed doesn't do what I want it to do. Tried to do something similar in a separate c++ experiment, and I do indeed need to bind it to a named variable. Thanks for the feedback!

File: src/workerd/util/duration-alert.h
Comment (Frederik-Baetens) [reply src/workerd/util/duration-alert.h]: Thanks! Marked it as noexcept(false), but I didn't include the UnwindDetector. I think that's fine?

File: src/workerd/util/duration-alert.h
Comment (jclee) [inline src/workerd/util/duration-alert.h:8]: I think the preferred style is to use quotes for relative includes (usually just files in the same directory) and angle brackets for absolute includes.

File: src/workerd/io/actor-cache.c++
Comment (jclee) [inline src/workerd/io/actor-cache.c++:2586]: I think typically we'd either declare the variable like
```
util::DurationAlert alert(clock, [...]);
```
or
```
auto alert = util::DurationAlert(clock, [...]);
```
just to avoid the repetition of writing the type twice.  Pretty sure the underscore variable name would be OK, but it's infrequently used.

File: src/workerd/util/duration-alert.h
Comment (jclee) [reply src/workerd/util/duration-alert.h]: Yeah, I think that's fine.  It seems like UnwindDetector is mostly used in places that either explicitly throw exceptions or could be reasonably anticipated to throw exceptions.

File: src/workerd/io/actor-cache.c++
Comment (joshthoward) [reply src/workerd/io/actor-cache.c++:2586]: For reference https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-list. I'd like to know generally whether we prefer list initialization or auto/assignment.

File: src/workerd/util/duration-alert.h
Comment (Frederik-Baetens) [reply src/workerd/util/duration-alert.h]: I added a unit test based on KJ_EXPECT_LOG :slightly_smiling_face: .

I wanted to test the negative case as well (no log emitted when the duration was not exceeded), but that'll require some KJ changes, and doesn't seem quite as critical to have a test for. I might add that after making a change to KJ

File: src/workerd/util/duration-alert-test.c++
Comment (jasnell) [inline src/workerd/util/duration-alert-test.c++:25]:
```suggestion
}  // namespace
}  // namespace workerd::util
```

File: src/workerd/io/actor-cache.c++
Comment (jasnell) [inline src/workerd/io/actor-cache.c++:2586]: Does this actually need the `kj::Duration(...)` around the calculation. Would `1 * kj::SECONDS` also yield a `kj::Duration`?

File: src/workerd/util/duration-alert.h
Comment (jasnell) [inline src/workerd/util/duration-alert.h:24]:
```suggestion
      clock(clock) {}
      
    KJ_DISALLOW_COPY_AND_MOVE(DurationAlert);
```

File: src/workerd/util/duration-alert.h
Comment (jasnell) [inline src/workerd/util/duration-alert.h:33]:
```suggestion
private:
```

File: src/workerd/util/duration-alert.h
Comment (jasnell) [inline src/workerd/util/duration-alert.h:29]:
```suggestion
      KJ_LOG(WARNING, kj::str("NOSENTRY ", logMessage), duration);
```

File: src/workerd/util/duration-alert.h
Comment (jclee) [reply src/workerd/util/duration-alert.h]: The test looks good, thanks!

File: src/workerd/util/duration-alert.h
Comment (jasnell) [reply src/workerd/util/duration-alert.h:29]: Actually, we already have a macro for handling the `NOSENTRY` bit here, don't we?

File: src/workerd/io/actor-cache.c++
Comment (jclee) [reply src/workerd/io/actor-cache.c++]: FWIW, in situations where either parens or braces would work, I've just stuck to using parens to match the existing kj-style code for consistency.  But I think this is not strictly enforced, and that there's some variation throughout the code base.

IIRC, there were situations for both parens and braces that were not totally covered by the other, and similarly for "almost always auto" style.  But with C++ making breaking-change improvements in recent years, maybe this has changed?

File: src/workerd/util/duration-alert.h
Comment (jclee) [reply src/workerd/util/duration-alert.h:29]: Heh... per earlier discussion, the current implementation of that macro requires a literal string as the first argument, so you'd have to do something like this anyway, to get it to work with the logMessage member variable.

https://github.com/cloudflare/workerd/pull/1693#discussion_r1495678123

File: src/workerd/io/actor-cache.c++
Comment (Frederik-Baetens) [reply src/workerd/io/actor-cache.c++:2586]: That does indeed also work, thanks!

File: src/workerd/util/duration-alert.h
Comment (Frederik-Baetens) [reply src/workerd/util/duration-alert.h:24]: Added this to the PR, but what's the rationale behind this?

Comment (Frederik-Baetens) [reply src/workerd/util/duration-alert.h:24]: I suppose having a move constructor for a raii object like this would lead to confusing behavior, as it'd be weird to track what's actually being timed. Copying probably doesn't make sense either for similar reasons, since you'd always want to re-initialize the start time rather than blindly copying.

File: src/workerd/util/duration-alert.h
Comment (justin-mp) [inline src/workerd/util/duration-alert.h:18]: Optional, non-blocking, nit: consider changing the name of this to DurationLogger or LongDurationLogger since that better describes what the class does.  (Alerts mean different things.)

**Review by justin-mp — CHANGES_REQUESTED** [review]

File: src/workerd/util/duration-alert.h
Comment (Frederik-Baetens) [reply src/workerd/util/duration-alert.h]: Thanks Justin! I've incorporated this, and also renamed it to "durationExceededLogger". I haven't gone as far as taking an operation description, since this might be useful in places where you're not specifically timing operations.

File: src/workerd/io/actor-cache.c++
Comment (justin-mp) [inline src/workerd/io/actor-cache.c++:2586]: nit: in this file, `1 * kj::SECONDS` is much easier to read than `kj::SECONDS * 1`.

**Review by justin-mp — APPROVED** [review]

**Review by jasnell — APPROVED** [review]

**Pull request merged**
